### ISSUE TITLE
Add support for Experiment tracking in Model Registry

### ISSUE THREAD

**Issue by dhirajsb** [kubeflow/model-registry#1224]
**Is your feature request related to a problem? Please describe.**
Model Registry currently only handles models and versions. ML/AI development starts with experiments and runs. Model Registry API MUST add support for creating and tracking the following:
* Experiments
* Runs within Experiments
* DataSets that are part of input data set
* Metrics and metric history associated with runs
* Parameters associated with runs

The above artifact data should also be accessible in Model Versions. 

**Describe the solution you'd like**
Add schema types for the Contexts (Experiment and Run), and Artifacts(DataSet, Metric, Parameter). All of this is possible to be supported and implemented in the current MLMD schema based architecture and REST API. 

**Describe alternatives you've considered**
Adding an mlflow compliant gRPC/RESTish API. This alternative was rejective due to the cognitive complexity introduced by such a bipolar API scheme with the existing model registry REST API and a partial gRPC API. 
MLflow compatibility is going to be addressed with another feature enhancement next. 

**Additional context**

**Comment by dhirajsb** [comment]
FYI, POC work for this has already stared at https://github.com/dhirajsb/model-registry-kfp/tree/feat/experiment-tracking-api

**Comment by dhirajsb** [comment]
# Model Registry API Changes: Experiment Tracking Feature

## Overview

This document outlines the comprehensive changes made to the Model Registry API in the `feat/experiment-tracking-api` branch compared to the main branch. The changes introduce experiment tracking capabilities to support MLflow integration.

## New Entities Added

### **Experiment**
- **Purpose**: Container for organizing multiple experiment runs
- **Key Properties**:
  - `name` (required): Unique identifier for the experiment
  - `owner`: Experiment owner ID or name
  - `state`: `LIVE` or `ARCHIVED`
- **Relationships**: Has many `ExperimentRun` children

### **ExperimentRun** 
- **Purpose**: Represents individual experiment execution instances
- **Key Properties**:
  - `experimentId` (required): Parent experiment reference
  - `startTimeSinceEpoch`/`endTimeSinceEpoch`: Execution timing
  - `status`: `RUNNING`, `SCHEDULED`, `FINISHED`, `FAILED`, `KILLED`
  - `state`: `LIVE` or `ARCHIVED`
  - `owner`: Run owner ID or name
- **Relationships**: Belongs to `Experiment`, has many `Artifact` children

### **DataSet** (new Artifact type)
- **Purpose**: Training/test data artifact representation
- **Key Properties**:
  - `digest`: Unique hash for dataset content
  - `sourceType`: Data source type (s3, hdfs, local, database)
  - `source`: Location/connection string
  - `schema`: JSON schema description
  - `profile`: Statistical summary
  - `uri`: Physical dataset location

### **Metric** (new Artifact type)
- **Purpose**: Numerical measurements from model training/evaluation
- **Key Properties**:
  - `name`: Metric identifier (accuracy, loss, f1_score)
  - `value`: Numeric measurement (double)
  - `timestamp`: Recording time (milliseconds)
  - `step`: Step number for multi-step metrics
- **Use Cases**: Training metrics, evaluation scores, performance monitoring

### **Parameter** (new Artifact type)
- **Purpose**: Configuration parameters used in training/execution
- **Key Properties**:
  - `name`: Parameter identifier (learning_rate, batch_size, epochs)
  - `value`: Parameter value (string representation)
  - `parameterType`: `string`, `number`, `boolean`, `object`
- **Use Cases**: Hyperparameters, configuration settings

## New Endpoints Added

### **Experiment Management**
- `GET /api/model_registry/v1alpha3/experiment` - Find experiment by name/externalId
- `GET /api/model_registry/v1alpha3/experiments` - List all experiments
- `POST /api/model_registry/v1alpha3/experiments` - Create new experiment
- `GET /api/model_registry/v1alpha3/experiments/{experimentId}` - Get experiment details
- `PATCH /api/model_registry/v1alpha3/experiments/{experimentId}` - Update experiment
- `GET /api/model_registry/v1alpha3/experiments/{experimentId}/experiment_runs` - List experiment's runs
- `POST /api/model_registry/v1alpha3/experiments/{experimentId}/experiment_runs` - Create run in experiment

### **ExperimentRun Management**
- `GET /api/model_registry/v1alpha3/experiment_run` - Find run by name/externalId/parentResourceId
- `GET /api/model_registry/v1alpha3/experiment_runs` - List all experiment runs
- `POST /api/model_registry/v1alpha3/experiment_runs` - Create new experiment run
- `GET /api/model_registry/v1alpha3/experiment_runs/{experimentrunId}` - Get run details
- `PATCH /api/model_registry/v1alpha3/experiment_runs/{experimentrunId}` - Update experiment run
- `GET/POST /api/model_registry/v1alpha3/experiment_runs/{experimentrunId}/artifacts` - Manage run artifacts
- `GET /api/model_registry/v1alpha3/experiment_runs/{experimentrunId}/metric_history` - Get metric history with step filtering

## Enhanced Existing Endpoints

The following endpoints received new filtering capabilities:

### **Added `filterQuery` Parameter**
- `GET /inference_services` - SQL-like filtering for inference services
- `GET /model_versions` - Enhanced model version filtering
- `GET /model_artifacts` - Enhanced model artifact filtering
- `GET /registered_models` - Enhanced registered model filtering
- `GET /registered_models/{id}/versions` - Enhanced version filtering
- `GET /serving_environments` - Enhanced serving environment filtering
- `GET /serving_environments/{id}/inference_services` - Enhanced nested resource filtering

### **Added `artifactType` Parameter**
- `GET /model_versions/{id}/artifacts` - Filter artifacts by type (model-artifact, doc-artifact, dataset-artifact, metric, parameter)

## New Query Parameters

### **`filterQuery`**
- **Purpose**: SQL-like filtering with rich capabilities
- **Supported Operators**: `=`, `!=`, `<>`, `>`, `<`, `>=`, `<=`, `LIKE`, `ILIKE`, `IN`, `AND`, `OR`
- **Data Types**: Strings, numbers, booleans with automatic type inference
- **Property Access**: Standard properties, custom properties, type-specific access
- **Examples**:
  - `name = "my-model"`
  - `accuracy > 0.95`
  - `name LIKE "%tensorflow%"`
  - `framework.string_value = "pytorch"`

### **`artifactType`**
- **Purpose**: Filter artifacts by specific types
- **Supported Values**: `model-artifact`, `doc-artifact`, `dataset-artifact`, `metric`, `parameter`

### **`stepIds`**
- **Purpose**: Filter metrics by specific step numbers
- **Format**: Comma-separated list of step IDs
- **Example**: `"1,2,3"`

## Additional Enhancements

### **Extended Artifact System**
- **Discriminator Enhancement**: Extended artifact discriminator to support all 5 artifact types
- **Type Safety**: Added `ArtifactTypeQueryParam` enum for consistent type filtering

### **Improved Ordering**
- **New OrderByField**: Added `NAME` to the `OrderByField` enum alongside existing `CREATE_TIME`, `LAST_UPDATE_TIME`, `ID`

### **Enhanced Response Links**
- **New Links**: Added comprehensive response links for experiment and experiment run entities
- **Navigation**: Improved API discoverability with proper HATEOAS links

## MLflow Integration Support

This API extension provides the foundation for comprehensive MLflow integration, supporting:
- **MLflow Experiments** → Model Registry `Experiment` entities
- **MLflow Runs** → Model Registry `ExperimentRun` entities  
- **MLflow Metrics** → Model Registry `Metric` artifacts
- **MLflow Parameters** → Model Registry `Parameter` artifacts
- **MLflow Datasets** → Model Registry `DataSet` artifacts

The enhanced filtering capabilities (`filterQuery`) enable efficient querying patterns required by MLflow's tracking store interface, while the new artifact types provide complete coverage of MLflow's metadata model.

## Summary

The experiment tracking feature adds significant capabilities to the Model Registry API:

- **5 new entity types** (Experiment, ExperimentRun, DataSet, Metric, Parameter)
- **15 new endpoints** for experiment and experiment run management
- **Enhanced filtering** across 8 existing endpoints
- **3 new query parameters** for improved data retrieval
- **Comprehensive MLflow support** enabling full tracking store integration

These changes maintain backward compatibility while extending the API to support modern ML experiment tracking workflows. 

The diagrams below shows the new entities added to the API:

<img width="7840" height="6210" alt="Image" src="https://github.com/user-attachments/assets/cb75c02e-36f6-4e81-b6d7-eed4ad145b95" />

The diagram below shows the new API endpoints:

<img width="7840" height="1990" alt="Image" src="https://github.com/user-attachments/assets/973c5c2f-4c2a-42da-aad8-35356aa433aa" />

**Comment by tarilabs** [comment]
thank you so much for this detailed proposition @dhirajsb !! 👏 

I have some minor comments based on previous discussions.

## New Entities Added

### Metric (new Artifact type)
I was under the impression you pointed this could be used with your modifications also for LMEvals and similar type of results when associated to a ModelVersion. I understood from you this was possible with a `uri` field also for this entity, but I don't see it in this report. Can you clarify, please?

Additionally, what happens when an "evaluation scores" listed in the use case is non-numerical?
I was under the impression here we wanted to support more than floating-points for the `value`s.

## Enhanced Existing Endpoints
Can you expand on what "Enhanced ... filtering" mentioned in this section entails, please?
Is that the `filterQuery` parameter listed in the following section?

**Comment by dhirajsb** [comment]
> 
> ## New Entities Added
> ### Metric (new Artifact type)
> I was under the impression you pointed this could be used with your modifications also for LMEvals and similar type of results when associated to a ModelVersion. I understood from you this was possible with a `uri` field also for this entity, but I don't see it in this report. Can you clarify, please?
> 

Since Metrics are Artifact types, we can also add them to ModelVersions (not shown in the diagrams). Same for paraameters. So, evals run after registering a version can still add parameters and metrics to modelversions. Metrics only have a float64 value in them right now, but they can also have customProperties that can contain more information about where the metrics came from. 

I didn't expose uri field in a metric or parameter because they have the actual artifact data, instead of being pointers to metadata like other artifacts. Whereas, uri in other artifact types points to a remote location with the actual data. 

Do you think a customProperty with reference information about the metric/parameter meets the eval requirement, or do you think uri is required? 

> Additionally, what happens when an "evaluation scores" listed in the use case is non-numerical? I was under the impression here we wanted to support more than floating-points for the `value`s.
> 

This distinction comes from MLflow, and I have seen a lot of other registries do it similarly. A Metric is a measurable quantity, hence they generalized on float64. If an eval or any other process wants to associate a non-numeric artifact, a parameter might be what's needed. Essentially, in MLFlow and other frameworks like [WeightsAndBiases](https://wandb.ai/site/), a bunch of metrics can be thrown in a chart and graphed. 

> ## Enhanced Existing Endpoints
> Can you expand on what "Enhanced ... filtering" mentioned in this section entails, please? Is that the `filterQuery` parameter listed in the following section?

Yes, it's referring to the filterQuery parameter. It's _enhanced_ because it supports SQL style WHERE clause for entity properties and customProperties.

**Comment by tarilabs** [comment]
Thanks @dhirajsb for the overall feedback;

> Do you think a customProperty with reference information about the metric/parameter meets the eval requirement, or do you think uri is required?

I think this is a good way of proceeding, and we can eventually consider surfacing it as a field if we see traction/usage. Do you concur with it being a possible future enhancement/refactor?

**Comment by dhirajsb** [comment]
> Thanks [@dhirajsb](https://github.com/dhirajsb) for the overall feedback;
> 
> > Do you think a customProperty with reference information about the metric/parameter meets the eval requirement, or do you think uri is required?
> 
> I think this is a good way of proceeding, and we can eventually consider surfacing it as a field if we see traction/usage. Do you concur with it being a possible future enhancement/refactor?

I Agree. We could introduce `source*` properties like we did for model artifacts.

**Comment by dhirajsb** [comment]
@tarilabs added tests to confirm that metrics and parameters can have descriptions as well. 
But, imho, it's not ideal if the same boilerplate description is copied into thousands of metric rows in the DB. It's better to have the description into an i18n resource bundle in the UI instead.

**Comment by kahowell** [comment]
@dhirajsb, I am wondering if we can simplify the `filterQuery` stuff. Do we actually need that level of dynamism? Can we instead use an explicit list of fields that we support filtering by, using a pattern of `$field_$comparison` (where just `$field` implies equality comparison), with an example of `?name=my-model&accuracy_gt=0.95&name_contains=tensorflow&framework=pytorch`. Having an explicit list will let us better optimize query performance and make it simpler (IMO) to use the API.

**Comment by dhirajsb** [comment]
@kahowell MLflow has support for [complex queries](https://mlflow.org/docs/1.23.1/search-syntax.html). So does [ml-metadata](https://github.com/google/ml-metadata/blob/d3a70e909463a231d367270c9f3ef89cf39a1df4/ml_metadata/proto/metadata_store.proto#L909), which is used in VertexAI. Also weights and biases allows users to build [complex queries](https://docs.wandb.ai/guides/app/features/panels/query-panels/#create-a-query-panel) with joins, etc. using UI. 

If we want to meet the data analysis needs of data scientists that have used or seen these other platforms, we have to be able to support complex queries. 

The MLflow plugin that builds on top of this server side work also needs this functionality to support mapping MLflow queries to MR.
